<a href="https://colab.research.google.com/github/piaseckazaneta/Python/blob/main/Zaneta_Piasecka_Mini_projekt_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 0. WSTĘP

Cel mini-projektu: Integracja rozproszonych danych relacyjnych NOAA (National Oceanic and Atmospheric Administration) i stworzenie czystego, ujednoliconego zbioru danych (tzw. Master Dataset) o ekstremalnych zjawiskach pogodowych, gotowego do analizy statystycznej i przestrzennej (GIS).

Z uwagi na to, że mini-projekt ma na celu zebranie i przygotowanie danych do analiz, w przedstawionym mini-projekcie nie ma wykresów ani wizualizacji. Będą one zaprezentowane w projekcie głównym.

Powyższe działania będą częścią projektu głównego, którego celem będzie identyfikacja profilu ryzyka ekstremalnych zjawisk pogodowych w USA poprzez analizę ich siły, rozkładu przestrzennego, generowanych strat ekonomicznych oraz wpływu na bezpieczeństwo ludności.

W projekcie głównym będą przeprowadzone:
- analiza czasowa (kiedy?),
- analiza finanansowa (ile wynoszą straty finansowe?),
- analiza ofiar zjawisk (ile osób, jaka płeć?)
- analiza przestrzenna (gdzie?)

Wyniki powyższych analiz będą przedstawione na wykresach i mapach.

Źródło danych: NOAA.


## 1. IMPORT BIBLIOTEK

In [ ]:
import pandas as pd

from google.colab import drive

drive.mount("/content/drive")



Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## 2. WCZYTANIE DANYCH

Dane pobierane są z dysku google, przez co nie ma potrzeby wczytywania ich przy każdym uruchomieniu kodu.



In [ ]:
# Ścieżki do dysku google
storm_events_details = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_details_2025.csv"
storm_events_locations = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_locations_2025.csv"
storm_events_fatalities = "/content/drive/MyDrive/Projekt_Python_2026/StormEvents_fatalities_2025.csv"

# Wczytanie danych
df_details = pd.read_csv(storm_events_details)
df_locations = pd.read_csv(storm_events_locations)
df_fatalities = pd.read_csv(storm_events_fatalities)


## 3. INSPEKCJA DANYCH

### 3.1. Kontrola struktury danych

In [ ]:
# Sprawdzam 5 pierwszych wierszy tabeli
df_details.head()

,BEGIN_YEARMONTH,BEGIN_DAY,BEGIN_TIME,END_YEARMONTH,END_DAY,END_TIME,EPISODE_ID,EVENT_ID,STATE,STATE_FIPS,...,END_RANGE,END_AZIMUTH,END_LOCATION,BEGIN_LAT,BEGIN_LON,END_LAT,END_LON,EPISODE_NARRATIVE,EVENT_NARRATIVE,DATA_SOURCE
0,202503,31,1104,202503,31,1106,201366,1252415,GEORGIA,13,...,2.22,W,TYUS,33.4757,-85.238,33.4757,-85.238,A cold-front initiated a line of thunderstorms...,Tree down at the intersection of highway 5 and...,CSV
1,202503,30,1552,202503,30,1555,200337,1241136,MICHIGAN,26,...,1.47,NNE,EDWARDSBURG,41.7900,-86.100,41.8200,-86.070,A cold front pushed into the area during the a...,A brief EF-1 tornado was confirmed in Edwardsb...,CSV
2,202501,5,1800,202501,6,2227,197733,1222851,VIRGINIA,51,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure tracked across souther...,NaN,CSV
3,202501,3,1300,202501,3,1900,197761,1223112,MARYLAND,24,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure moved off into New Eng...,NaN,CSV
4,202501,3,1300,202501,3,1900,197761,1223113,MARYLAND,24,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,An area of low pressure moved off into New Eng...,NaN,CSV


In [ ]:
# Sprawdzam 5 pierwszych wierszy tabeli
df_fatalities.head()

,FAT_YEARMONTH,FAT_DAY,FAT_TIME,FATALITY_ID,EVENT_ID,FATALITY_TYPE,FATALITY_DATE,FATALITY_AGE,FATALITY_SEX,FATALITY_LOCATION
0,202512,4,0,60389,1296292,I,12/04/2025 00:00:00,40.0,M,Vehicle/Towed Trailer
1,202512,1,0,60391,1296789,I,12/01/2025 00:00:00,35.0,F,Boating
2,202512,3,0,60393,1297305,D,12/03/2025 00:00:00,32.0,M,In Water
3,202512,25,0,60553,1298352,I,12/25/2025 00:00:00,75.0,F,Vehicle/Towed Trailer
4,202512,26,0,60562,1298509,D,12/26/2025 00:00:00,44.0,M,In Water


In [ ]:
# Sprawdzam 5 pierwszych wierszy tabeli
df_locations.head()

,YEARMONTH,EPISODE_ID,EVENT_ID,LOCATION_INDEX,RANGE,AZIMUTH,LOCATION,LATITUDE,LONGITUDE,LAT2,LON2
0,202508,204641,1287936,2,3.54,SSE,SINKING COVE,35.0015,-85.9601,3590,8557606
1,202508,203789,1269348,1,2.90,E,BLEASES XRDS,34.0440,-81.7699,342640,8146194
2,202508,203789,1269349,1,0.80,WSW,DOUGLAS,34.6637,-80.8118,3439822,8048708
3,202508,204597,1288676,4,1.45,WSW,HUNTSVILLE,34.7237,-86.6444,3443422,8638664
4,202508,204597,1288677,1,1.79,NW,FT PAYNE ISBELL ARPT,34.4900,-85.7400,3429400,8544400


**Komentarz**:
Tabele wyglądają na uporządkowane.
Poprawnie wyglądają formaty dat, skrótowe oznaczenia kierunków świata.

X Nazwy kolumn pisane wielkimi literami - Należy je zamienić na małe litery.

Wyświetlenie wszystkich atrybutów tabeli df_details, ponieważ z uwagi na ich dużą liczbę, head() nie wyświetla wszystkich.

In [ ]:
# Wyświetlam wszystkie nazwy kolumn z tabeli
df_details.columns.to_list()

['BEGIN_YEARMONTH',
 'BEGIN_DAY',
 'BEGIN_TIME',
 'END_YEARMONTH',
 'END_DAY',
 'END_TIME',
 'EPISODE_ID',
 'EVENT_ID',
 'STATE',
 'STATE_FIPS',
 'YEAR',
 'MONTH_NAME',
 'EVENT_TYPE',
 'CZ_TYPE',
 'CZ_FIPS',
 'CZ_NAME',
 'WFO',
 'BEGIN_DATE_TIME',
 'CZ_TIMEZONE',
 'END_DATE_TIME',
 'INJURIES_DIRECT',
 'INJURIES_INDIRECT',
 'DEATHS_DIRECT',
 'DEATHS_INDIRECT',
 'DAMAGE_PROPERTY',
 'DAMAGE_CROPS',
 'SOURCE',
 'MAGNITUDE',
 'MAGNITUDE_TYPE',
 'FLOOD_CAUSE',
 'CATEGORY',
 'TOR_F_SCALE',
 'TOR_LENGTH',
 'TOR_WIDTH',
 'TOR_OTHER_WFO',
 'TOR_OTHER_CZ_STATE',
 'TOR_OTHER_CZ_FIPS',
 'TOR_OTHER_CZ_NAME',
 'BEGIN_RANGE',
 'BEGIN_AZIMUTH',
 'BEGIN_LOCATION',
 'END_RANGE',
 'END_AZIMUTH',
 'END_LOCATION',
 'BEGIN_LAT',
 'BEGIN_LON',
 'END_LAT',
 'END_LON',
 'EPISODE_NARRATIVE',
 'EVENT_NARRATIVE',
 'DATA_SOURCE']

Komentarz:
.head() ucięło kolumny w df_details.

Prezentacja nazw kolumn w postaci listy umożliwia zobaczenie wszystkich kolumn w tabeli.

Sprawdzenie rozmiarów tabel i typów kolumn

In [ ]:
# Sprawdzam rozmiar tabeli i typy kolumn
df_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72241 entries, 0 to 72240
Data columns (total 51 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   BEGIN_YEARMONTH     72241 non-null  int64  
 1   BEGIN_DAY           72241 non-null  int64  
 2   BEGIN_TIME          72241 non-null  int64  
 3   END_YEARMONTH       72241 non-null  int64  
 4   END_DAY             72241 non-null  int64  
 5   END_TIME            72241 non-null  int64  
 6   EPISODE_ID          72241 non-null  int64  
 7   EVENT_ID            72241 non-null  int64  
 8   STATE               72241 non-null  object 
 9   STATE_FIPS          72241 non-null  int64  
 10  YEAR                72241 non-null  int64  
 11  MONTH_NAME          72241 non-null  object 
 12  EVENT_TYPE          72241 non-null  object 
 13  CZ_TYPE             72241 non-null  object 
 14  CZ_FIPS             72241 non-null  int64  
 15  CZ_NAME             72241 non-null  object 
 16  WFO 

Komentarz:

Użycie .info() służyło sprawdzeniu typów danych i potencjalnych braków danych dzięki kolumnie Non-Null Count.. Można odczytać też liczbę wierszy w tabeli.

Całkowita liczba wierszy: 72241.

X Daty ("date") są w formacie int64 - do konwersji.
X Kwoty ("damage") dotyczące zniszczeń są w formacie object - do konwersji.
X Skala tornada (tor_f_scale) jest w formacie object - do późniejszych obliczeń będzie potrzebny format liczbowy.

Braki w danych:
ok. 20 tys. z 72 tys. brakujących wartości w kolumnie "damage_crops" - do sprawdzenia (powinno być 0, czy zdarzenie nie dotyczy zniszczeń upraw?)
ok. 70 tys. brakujących wartości dla skali, długości i szerokości tornad (tor_f_scale, tor_length, tor_width) - do sprawdzenia, prawdopodobnie brakujące wiersze są innymi zdarzeniami niż tornado.

Reszta interesujących kolumn ma poprawny format i kompletną liczbę wartości.


In [ ]:
# Sprawdzam rozmiar tabeli i typy kolumn
df_fatalities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 895 entries, 0 to 894
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   FAT_YEARMONTH      895 non-null    int64  
 1   FAT_DAY            895 non-null    int64  
 2   FAT_TIME           895 non-null    int64  
 3   FATALITY_ID        895 non-null    int64  
 4   EVENT_ID           895 non-null    int64  
 5   FATALITY_TYPE      895 non-null    object 
 6   FATALITY_DATE      895 non-null    object 
 7   FATALITY_AGE       743 non-null    float64
 8   FATALITY_SEX       785 non-null    object 
 9   FATALITY_LOCATION  895 non-null    object 
dtypes: float64(1), int64(5), object(4)
memory usage: 70.1+ KB


Komentarz:

Całkowita liczba wierszy: 895.
Wszystkie kolumny mają poprawny format danych.
W "fatality_age" i "fatality_sex" ok. 100 brakujących wartości - do sprawdzenia.

In [ ]:
# Sprawdzam rozmiar tabeli i typy kolumn
df_locations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51870 entries, 0 to 51869
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   YEARMONTH       51870 non-null  int64  
 1   EPISODE_ID      51870 non-null  int64  
 2   EVENT_ID        51870 non-null  int64  
 3   LOCATION_INDEX  51870 non-null  int64  
 4   RANGE           51870 non-null  float64
 5   AZIMUTH         51870 non-null  object 
 6   LOCATION        51870 non-null  object 
 7   LATITUDE        51870 non-null  float64
 8   LONGITUDE       51870 non-null  float64
 9   LAT2            51870 non-null  int64  
 10  LON2            51870 non-null  int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 4.4+ MB


Komentarz:

Całkowita liczba wierszy: 51870.
Wszystkie kolumny mają poprawny format danych i 0 brakujących wartości.

### 3.2. Sprawdzenie braków w danych i duplikatów

In [ ]:
# Sprawdzam liczbę brakujących wartości w poszczególnych kolumnach
df_details.isnull().sum()

,0
BEGIN_YEARMONTH,0
BEGIN_DAY,0
BEGIN_TIME,0
END_YEARMONTH,0
END_DAY,0
END_TIME,0
EPISODE_ID,0
EVENT_ID,0
STATE,0
STATE_FIPS,0


In [ ]:
# Sprawdzam liczbę brakujących wartości w poszczególnych kolumnach
df_fatalities.isnull().sum()

,0
FAT_YEARMONTH,0
FAT_DAY,0
FAT_TIME,0
FATALITY_ID,0
EVENT_ID,0
FATALITY_TYPE,0
FATALITY_DATE,0
FATALITY_AGE,152
FATALITY_SEX,110
FATALITY_LOCATION,0


In [ ]:
# Sprawdzam liczbę brakujących wartości w poszczególnych kolumnach
df_locations.isnull().sum()

,0
YEARMONTH,0
EPISODE_ID,0
EVENT_ID,0
LOCATION_INDEX,0
RANGE,0
AZIMUTH,0
LOCATION,0
LATITUDE,0
LONGITUDE,0
LAT2,0


Sprawdzenie unikalności kolumny "EVENT_ID", gdyż będzie to kolumna klucz do integracji tabel.

In [ ]:
# Sprawdzam, czy kolumna "EVENT_ID" posiada unikalne wartości
df_details["EVENT_ID"].is_unique

True

Kolumna "EVENT_ID" głównej tabeli df_details posiada unikalne wartości

In [ ]:
# Sprawdzam, czy kolumna "EVENT_ID" posiada unikalne wartości, używając sposobu na sprawdzenie liczby duplikatów
df_details["EVENT_ID"].duplicated().sum()

np.int64(0)

Kolumna "EVENT_ID" nie ma duplikatów, zatem spełnia warunki klucza głównego (Primary Key) dla tabeli df_details

In [ ]:
# Sprawdzam liczbę duplikatów w tabeli z lokalizacją
df_locations["EVENT_ID"].duplicated().sum()

np.int64(20973)

Co ciekawe dla "EVENT_ID" w tabeli df_locations jest prawie 21 tys. duplikatów, co wcale nie musi oznaczać błędu. Dane zjawisko pogodowe zarejestrowane w tabeli df_details mogło przemieszczać się przez kilka lokalizacji.
Do sprawdzenia.

In [ ]:
# Sprawdzam liczbę duplikatów w tabeli z ofiarami
df_fatalities["EVENT_ID"].duplicated().sum()

np.int64(415)

EVENT_ID w tabeli df_fatalities również posiada duplikaty.
Prawdopodobnie jest tak, że w zdarzeniach zginęło wiele ofiar.

Jeśli chodzi o poziomy agregacji danych, to tabela df_details posiada dane jednostkowe - poziom pojedynczego zdarzenia pogodowego.

W tabeli df_locations jeden wiersz to jedna lokalizacja geograficzna.

W tabeli df_fatalities jeden wiersz to jedna ofiara zdarzenia.

### 3.3. Sprawdzenie miar statystyki opisowej

In [ ]:
# Sprawdzam miary statystyczne do ogólnego spojrzenia na dane, ewentualne wartości odstające
df_details.describe(include='all')

,BEGIN_YEARMONTH,BEGIN_DAY,BEGIN_TIME,END_YEARMONTH,END_DAY,END_TIME,EPISODE_ID,EVENT_ID,STATE,STATE_FIPS,...,END_RANGE,END_AZIMUTH,END_LOCATION,BEGIN_LAT,BEGIN_LON,END_LAT,END_LON,EPISODE_NARRATIVE,EVENT_NARRATIVE,DATA_SOURCE
count,72241.000000,72241.000000,72241.000000,72241.000000,72241.000000,72241.000000,72241.000000,7.224100e+04,72241,72241.000000,...,43917.000000,43917,43917,43917.000000,43917.000000,43917.000000,43917.000000,72241,61496,72241
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,68,NaN,...,NaN,16,16971,NaN,NaN,NaN,NaN,8898,53486,1
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,TEXAS,NaN,...,NaN,N,FLORENCE,NaN,NaN,NaN,NaN,A cluster of supercells developed during the l...,Maximum heat index of 105 degrees.,CSV
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5641,NaN,...,NaN,6807,48,NaN,NaN,NaN,NaN,412,447,72241
mean,202505.620451,15.113038,1281.282139,202505.620451,16.530488,1449.093368,203223.827826,1.265526e+06,NaN,33.584488,...,2.382360,NaN,NaN,37.430153,-90.057825,37.430663,-90.053952,NaN,NaN,NaN
std,2.992886,8.661094,659.785132,2.992886,8.681507,616.808726,3181.344483,2.307369e+04,NaN,19.114886,...,4.414841,NaN,NaN,5.205849,11.339800,5.206291,11.339300,NaN,NaN,NaN
min,202501.000000,1.000000,0.000000,202501.000000,1.000000,0.000000,197521.000000,1.221835e+06,NaN,1.000000,...,0.000000,NaN,NaN,-14.338000,-170.833900,-14.338000,-170.832600,NaN,NaN,NaN
25%,202503.000000,8.000000,835.000000,202503.000000,9.000000,1100.000000,200491.000000,1.245845e+06,NaN,20.000000,...,0.690000,NaN,NaN,34.328900,-97.150000,34.329800,-97.150000,NaN,NaN,NaN
50%,202505.000000,16.000000,1438.000000,202505.000000,17.000000,1600.000000,202803.000000,1.265417e+06,NaN,34.000000,...,1.380000,NaN,NaN,37.700000,-88.714700,37.702800,-88.710000,NaN,NaN,NaN
75%,202507.000000,21.000000,1800.000000,202507.000000,24.000000,1900.000000,205750.000000,1.284940e+06,NaN,47.000000,...,2.680000,NaN,NaN,40.640000,-81.730000,40.643500,-81.721300,NaN,NaN,NaN


In [ ]:
df_fatalities.describe(include='all')

,FAT_YEARMONTH,FAT_DAY,FAT_TIME,FATALITY_ID,EVENT_ID,FATALITY_TYPE,FATALITY_DATE,FATALITY_AGE,FATALITY_SEX,FATALITY_LOCATION
count,895.000000,895.000000,895.0,895.000000,8.950000e+02,895,895,743.000000,785,895
unique,NaN,NaN,NaN,NaN,NaN,2,224,NaN,2,16
top,NaN,NaN,NaN,NaN,NaN,D,07/04/2025 00:00:00,NaN,M,Outside/Open Areas
freq,NaN,NaN,NaN,NaN,NaN,632,124,NaN,501,185
mean,202505.707263,12.964246,0.0,59499.931844,1.270745e+06,NaN,NaN,49.893674,NaN,NaN
std,2.729264,8.340144,0.0,648.222758,2.260949e+04,NaN,NaN,22.699270,NaN,NaN
min,202501.000000,1.000000,0.0,58189.000000,1.222200e+06,NaN,NaN,1.000000,NaN,NaN
25%,202504.000000,5.000000,0.0,59031.500000,1.250102e+06,NaN,NaN,34.000000,NaN,NaN
50%,202506.000000,12.000000,0.0,59470.000000,1.278903e+06,NaN,NaN,53.000000,NaN,NaN
75%,202507.000000,19.000000,0.0,60007.500000,1.287451e+06,NaN,NaN,68.000000,NaN,NaN


In [ ]:
df_locations.describe(include='all')

,YEARMONTH,EPISODE_ID,EVENT_ID,LOCATION_INDEX,RANGE,AZIMUTH,LOCATION,LATITUDE,LONGITUDE,LAT2,LON2
count,51870.000000,51870.000000,5.187000e+04,51870.000000,51870.000000,51870,51870,51870.000000,51870.000000,5.187000e+04,5.187000e+04
unique,NaN,NaN,NaN,NaN,NaN,16,14902,NaN,NaN,NaN,NaN
top,NaN,NaN,NaN,NaN,NaN,N,CITADEL,NaN,NaN,NaN,NaN
freq,NaN,NaN,NaN,NaN,NaN,6452,113,NaN,NaN,NaN,NaN
mean,202506.248082,203722.326239,1.269588e+06,1.896221,2.544902,NaN,NaN,37.208435,-90.528872,3.142227e+06,7.574091e+06
std,2.124867,2528.497539,1.852537e+04,1.325653,5.216425,NaN,NaN,5.632769,13.600623,1.348388e+06,3.387846e+06
min,202501.000000,197604.000000,1.222429e+06,1.000000,0.000000,NaN,NaN,-14.342000,-170.833900,-1.420520e+06,-1.711839e+07
25%,202505.000000,201933.000000,1.256884e+06,1.000000,0.750000,NaN,NaN,33.993850,-97.830000,3.124194e+06,7.630994e+06
50%,202506.000000,203512.500000,1.270082e+06,1.000000,1.430000,NaN,NaN,37.668950,-88.124600,3.626169e+06,8.421135e+06
75%,202507.000000,205602.000000,1.284595e+06,3.000000,2.730000,NaN,NaN,40.297125,-80.270000,3.935640e+06,9.645600e+06


Nie zauważyłam żadnych nieprawidłowości po wyświetleniu miar statystycznych.

## 4. SELEKCJA ATRYBUTÓW

Ten krok ma na celu odcięcie szumu informacyjnego i wyselekcjonowanie tylko tych kolumn, które będą potrzebne do późniejszych analiz.

### 4.1. Zmiana nazw kolumn w df-ach na małe litery

In [ ]:
df_details.columns = df_details.columns.str.lower()
df_fatalities.columns = df_fatalities.columns.str.lower()
df_locations.columns = df_locations.columns.str.lower()

### 4.2. Selekcja zmiennych do analiz

In [ ]:
df_details_analysis = df_details[["event_id", "state", "event_type", "begin_date_time", "damage_property", "damage_crops", "tor_f_scale", "tor_length", "tor_width"]].copy()

In [ ]:
df_fatalities_analysis = df_fatalities[["event_id", "fatality_age", "fatality_sex", "fatality_location"]].copy()

In [ ]:
df_locations_analysis = df_locations[["event_id", "latitude", "longitude"]].copy()

Sprawdzenie wartości w wyselekcjonowanych kolumnach

In [ ]:
df_details_analysis.head()

,event_id,state,event_type,begin_date_time,damage_property,damage_crops,tor_f_scale,tor_length,tor_width
0,1252415,GEORGIA,Thunderstorm Wind,31-MAR-25 11:04:00,1.00K,NaN,NaN,NaN,NaN
1,1241136,MICHIGAN,Tornado,30-MAR-25 15:52:00,100.00K,0.00K,EF1,2.59,100.0
2,1222851,VIRGINIA,Winter Storm,05-JAN-25 18:00:00,NaN,NaN,NaN,NaN,NaN
3,1223112,MARYLAND,Winter Weather,03-JAN-25 13:00:00,NaN,NaN,NaN,NaN,NaN
4,1223113,MARYLAND,Winter Weather,03-JAN-25 13:00:00,NaN,NaN,NaN,NaN,NaN


Wnioski dotyczące wartości w kolumnach tabeli:
- wartości zniszczeń ("damage property" i "damage crops" są tekstowe -> należy je zamienić na wartości numeryczne w celu późniejszego oszacowania zniszczeń
- wartości w kolumnie "begin_date_time" zawierają date i czas -> należy te informacje rozdzielić na osobne kolumny, aby móc później określić sezonowość (miesiąc) i porę dnia (godzina) poszczególnych zjawisk
- "tor_f_scale" - Standaryzacja skali tornad: Oczyszczenie kolumny tor_f_scale (wyciągnięcie samej cyfry ze stringów typu "EF2").


## 5. KONWERSJA DANYCH

Cel: Naprawa i ujednolicenie formatów

### 5.1. Konwersja dat

Konwersja kolumny "begin_date_time" z df_details_analysis

In [ ]:
# Konwertuję kolumnę z object na datetime
df_details_analysis["begin_date_time"] = pd.to_datetime(df_details_analysis["begin_date_time"])

/tmp/ipykernel_8431/292393041.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_details_analysis["begin_date_time"] = pd.to_datetime(df_details_analysis["begin_date_time"])


Wyciągnięcie z daty nowych kolumn: month (do analizy sezonowej) oraz hour (do analizy porowej)

In [ ]:
# Wyświetlam kolumnę z datą i godziną po konwersji, aby sprawdzić jak wygląda
df_details_analysis["begin_date_time"]

,begin_date_time
0,2025-03-31 11:04:00
1,2025-03-30 15:52:00
2,2025-01-05 18:00:00
3,2025-01-03 13:00:00
4,2025-01-03 13:00:00
...,...
72236,2025-09-16 14:53:00
72237,2025-09-16 14:54:00
72238,2025-09-09 16:58:00
72239,2025-09-09 19:32:00


In [ ]:
# Tworzę kolumnę z numerem miesiąca zdarzenia
df_details_analysis["month"] = df_details_analysis["begin_date_time"].dt.month

In [ ]:
print(df_details_analysis["month"])

0        3
1        3
2        1
3        1
4        1
        ..
72236    9
72237    9
72238    9
72239    9
72240    9
Name: month, Length: 72241, dtype: int32


In [ ]:
# Tworzę kolumnę z pełną godziną zdarzenia
df_details_analysis["hour"] = df_details_analysis["begin_date_time"].dt.hour

In [ ]:
print(df_details_analysis["hour"])

0        11
1        15
2        18
3        13
4        13
         ..
72236    14
72237    14
72238    16
72239    19
72240    18
Name: hour, Length: 72241, dtype: int32


### 5.2. Konwersja wartości zniszczeń

Funkcja zamieniająca tekstowe wartości zniszczeń (np. 10.00K, 2.50M) na czyste liczby typu float ($10000$, $2500000$) dla kolumn "damage_crops" i "damage_property" w df_details_analysis

In [ ]:
# Wyświetlam unikalne wartości strat
df_details_analysis["damage_property"].unique()

array(['1.00K', '100.00K', nan, '0.00K', '10.00K', '50.00K', '25.00K',
       '5.00K', '15.00K', '20.00K', '9.00K', '3.00K', '2.00K', '30.00K',
       '4.00K', '6.00K', '0.50K', '2.50K', '1.50K', '1.10K', '18.50M',
       '17.80M', '3.40M', '8.00K', '55.00K', '75.00K', '3.50K', '40.00K',
       '35.00K', '0.60K', '1.00M', '5.00M', '2.50M', '0.20K', '0.30K',
       '200.00K', '250.00K', '150.00K', '260.00K', '700.00K', '10.50K',
       '600.00K', '900.00K', '7.00K', '12.00K', '300.00K', '11.00K',
       '2.25K', '49.50K', '0.10K', '500.00K', '160.00K', '70.00K',
       '60.00K', '45.00K', '7.50K', '0.25K', '80.00K', '13.00K', '0.75K',
       '140.00K', '3.90M', '90.00K', '350.00K', '14.00K', '1.20M',
       '12.50K', '27.00K', '0.01K', '0.00M', '18.00M', '33.00M', '56.00M',
       '120.00K', '0.40K', '125.00K', '5000.00K', '870.00K', '800.00K',
       '19.00K', '1.50M', '148.00K', '268.00K', '222.00K', '825.00K',
       '2.08M', '1.20K', '1.60K', '22.50K', '750.00K', '215.00K',
       '

In [ ]:
# Tworzę funkcję konwertującą wartości strat,
# aby pozbyć się liter i przedstawić pełną liczbę np. 1 000 000 a nie 1 M
def damage_costs(damage_value):
  if pd.isna(damage_value):
    return 0
  elif "K" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000
    return damage_value
  elif "M" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000000
    return damage_value
  elif "B" in damage_value:
    damage_value = damage_value[:-1]
    damage_value = float(damage_value) * 1000000000
    return damage_value
  else:
    float(damage_value)

In [ ]:
# Implementuję funkcję dla kolumny ze stratami w nieruchomościach
df_details_analysis["damage_property"] = df_details_analysis["damage_property"].apply(damage_costs)

In [ ]:
# Sprawdzam wyniki działania funkcji
print(df_details_analysis["damage_property"])

0          1000.0
1        100000.0
2             0.0
3             0.0
4             0.0
           ...   
72236         0.0
72237         0.0
72238         0.0
72239         0.0
72240      3000.0
Name: damage_property, Length: 72241, dtype: float64


In [ ]:
# Implementuję funkcję dla kolumny ze stratami w rolnictwie
df_details_analysis["damage_crops"] = df_details_analysis["damage_crops"].apply(damage_costs)

In [ ]:
# Sprawdzam wyniki po zastosowaniu funkcji
print(df_details_analysis["damage_crops"])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
72236    0.0
72237    0.0
72238    0.0
72239    0.0
72240    0.0
Name: damage_crops, Length: 72241, dtype: float64


### 5.3. Konwersja wartości kolumny skali tornad

Wyciągnięcie z kolumny "tor_f_scale" samej liczby (wartość wyrażona jest w skali Fujita, np. EF1). Wartość liczbowa będzie niezbędna do obliczeń zależności powierzchni zniszczeń od wzrostu siły tornada.

Sprawdzenie unikalnych wartości występujących w kolumnie "tor_f_scale"

In [ ]:
# Sprawdzam unikalne wartości skali tornad
df_details_analysis["tor_f_scale"].unique()

array([nan, 'EF1', 'EF0', 'EF3', 'EFU', 'EF2', 'EF4', 'EF5'], dtype=object)

Funkcja wyodrębniająca cyfrę z wartości w kolumnie "tor_f_scale".
'EFU' jest oznaczeniem tornada, które przechodziło przez pole i nie spowodowało uszkodzeń budynków, zatem nie można mu przypisać oznaczenia w skali 1-5.

In [ ]:
# Funkcja wyodrębniająca cyfry ze skali tornad
def scale_number(scale):
  if pd.isna(scale):
    return 'NaN'
  elif "EFU" in scale:
    return 'NaN'
  elif "EF" in scale:
    scale = scale[2:]
    return scale
  else:
    return scale

Zastosowanie napisanej funkcji dla kolumny "tor_f_scale" i utworzenie nowej kolumny z samą cyfrą.

In [ ]:
# Implementuję funkcję dla kolumny ze skalą tornad
df_details_analysis["tor_f_scale_number"] = df_details_analysis["tor_f_scale"].apply(scale_number)

Sprawdzenie unikalnych wartości występujących w nowej kolumnie

In [ ]:
# Sprawdzam wyniki po zastosowaniu funkcji
df_details_analysis["tor_f_scale_number"].unique()

array(['NaN', '1', '0', '3', '2', '4', '5'], dtype=object)

Zamiana wartości na float.

In [ ]:
# Konwertuję wartości z object na float
df_details_analysis["tor_f_scale_number"] = (df_details_analysis["tor_f_scale_number"]).astype(float)

## 6. INTEGRACJA DANYCH

### 6.1. Agregacja tabel z lokalizacją "df_location" i z ofiarami "df_fatalities"

Tabela df_fatalities_analysis oraz df_locations_analysis mają duplikaty "event_id". Gdyby połączyć je bezpośrednio, dane o stratach finansowych (damage) byłyby przekłamane po połaczeniu tabel, bo np. dla 1 zdarzenia przypisane byłyby 3 ofiary oraz 4 lokalizacje (wiersze zostałyby powielone).

Poniżej przygotowuję zagregowane wersje tych tabel na poziomie pojedynczego event_id:

In [ ]:
# Zliczamy liczbę ofiar dla każdego zdarzenia
df_fatalities_agg = df_fatalities_analysis.groupby('event_id').size().reset_index(name='fatalities_count')

In [ ]:
# Dla lokalizacji: bierzemy pierwszy zarejestrowany punkt geograficzny dla każdego zdarzenia
df_locations_agg = df_locations_analysis.drop_duplicates(subset=['event_id'])

In [ ]:
# Sprawdzenie wyglądu tabel po agregacji

print(df_fatalities_agg.head(10))
print(df_locations_agg.head(10))

   event_id  fatalities_count
0   1222200                 1
1   1222740                 1
2   1224881                 1
3   1225049                 1
4   1225562                 1
5   1225583                 1
6   1226112                 1
7   1226791                 2
8   1227246                 2
9   1227504                 1
    event_id  latitude  longitude
0    1287936   35.0015   -85.9601
1    1269348   34.0440   -81.7699
2    1269349   34.6637   -80.8118
3    1288676   34.7237   -86.6444
4    1288677   34.4900   -85.7400
8    1288678   34.4900   -85.7400
12   1287411   39.5200   -85.7600
13   1275406   34.0932  -102.1300
14   1275460   33.6700  -101.8200
15   1275461   33.5634  -102.4182


### 6.2. Łączenie tabel

In [ ]:
# Łączę tabelę główną (df_details_analysis) z tabelą z lokalizacjami po agregacji (df_locations_agg)
df_full = pd.merge(df_details_analysis, df_locations_agg, on='event_id', how='left')

In [ ]:
# Łączę powyższe z tabelą z ofiarami po agregacji (df_fatalities_agg)
df_full = pd.merge(df_full, df_fatalities_agg, on='event_id', how='left')

In [ ]:
# Sprawdzam rozmiar końcowej bazy – liczba wierszy musi być DOKŁADNIE taka sama jak w df_details_analysis
print("Rozmiar bazy Details:", df_details_analysis.shape)
print("Rozmiar bazy po połączeniu:", df_full.shape)

Rozmiar bazy Details: (72241, 12)
Rozmiar bazy po połączeniu: (72241, 15)


Komentarz: Liczba wierszy w obu tabelach (przed i po połączeniu) jest identyczna.

In [ ]:
# Szybki podgląd końcowego efektu
df_full.head()

,event_id,state,event_type,begin_date_time,damage_property,damage_crops,tor_f_scale,tor_length,tor_width,month,hour,tor_f_scale_number,latitude,longitude,fatalities_count
0,1252415,GEORGIA,Thunderstorm Wind,2025-03-31 11:04:00,1000.0,0.0,NaN,NaN,NaN,3,11,NaN,NaN,NaN,NaN
1,1241136,MICHIGAN,Tornado,2025-03-30 15:52:00,100000.0,0.0,EF1,2.59,100.0,3,15,1.0,NaN,NaN,NaN
2,1222851,VIRGINIA,Winter Storm,2025-01-05 18:00:00,0.0,0.0,NaN,NaN,NaN,1,18,NaN,NaN,NaN,NaN
3,1223112,MARYLAND,Winter Weather,2025-01-03 13:00:00,0.0,0.0,NaN,NaN,NaN,1,13,NaN,NaN,NaN,NaN
4,1223113,MARYLAND,Winter Weather,2025-01-03 13:00:00,0.0,0.0,NaN,NaN,NaN,1,13,NaN,NaN,NaN,NaN


## 7. WNIOSKI

Wynikiem mini-projektu jest kompletna do analizy statystycznej i przestrzennej tabela powstała z integracji 3 tabel: o zjawiskach pogodowych, ich lokalizacji i ofiarach tych zjawisk.

W jednym wierszu są informacje: Gdzie (Lat, Lon, Stan), Kiedy (czysta data), Co się stało (Tornado, Powódź), Ile kosztowało (czysta liczba w dolarach) oraz Kto ucierpiał (wiek, płeć, lokalizacja ofiary).